Testnet/NFTStore/NFTStore/NFTStore.cs

```diff
@@ -0,0 +1,155 @@
+    private void ClearSaleInfo(Address contract, ulong tokenId) => State.Clear($"SaleInfo:{contract}:{tokenId}");
+
+    public NFTStore(ISmartContractState state)
+        : base(state)
+    {
+        EnsureNotPayable();
+    }
+
+    /// <summary>
+    /// Ensure the token is approved on the non fungible contract for store contract as pre-condition.
+    /// </summary>
+    public void Sale(Address contract, ulong tokenId, ulong price)
+    {
+        EnsureNotPayable();
+
+        Assert(price > 0, "Price should be higher than zero.");
+
+        var tokenOwner = GetOwner(contract, tokenId);
+
+        Assert(tokenOwner != Address, "The token is already on sale.");
+
+        EnsureCallerCanOperate(contract, tokenOwner);
+
+        TransferToken(contract, tokenId, tokenOwner, Address);
+
+        SetSaleInfo(contract, tokenId, new SaleInfo { Price = price, Seller = tokenOwner });
+
+        Log(new TokenOnSaleLog { Contract = contract, TokenId = tokenId, Price = price, Seller = tokenOwner });
+    }
+
+    public void Buy(Address contract, ulong tokenId)
+    {
+        var saleInfo = GetSaleInfo(contract, tokenId);
+
+        EnsureTokenIsOnSale(saleInfo);
+
+        Assert(Message.Value == saleInfo.Price, "Transferred amount is not matching exact price of the token.");
+
+        SafeTransferToken(contract, tokenId, Address, Message.Sender);
+
+        ClearSaleInfo(contract, tokenId);
+
+        var result = Transfer(saleInfo.Seller, saleInfo.Price);
+
+        Assert(result.Success, "Transfer failed.");
+
+        Log(new TokenPurchasedLog { Contract = contract, TokenId = tokenId, Buyer = Message.Sender, Seller = saleInfo.Seller });
+    }
+
+    public void CancelSale(Address contract, ulong tokenId)
+    {
+        EnsureNotPayable();
+        var saleInfo = GetSaleInfo(contract, tokenId);
+
+        EnsureTokenIsOnSale(saleInfo);
+
+        EnsureCallerCanOperate(contract, saleInfo.Seller);
+
+        SafeTransferToken(contract, tokenId, Address, saleInfo.Seller);
+
+        ClearSaleInfo(contract, tokenId);
+
+        Log(new TokenSaleCanceledLog { Contract = contract, TokenId = tokenId, Seller = saleInfo.Seller });
+    }
+
+    private bool IsApprovedForAll(Address contract, Address tokenOwner)
+    {
+        var result = Call(contract, 0, "IsApprovedForAll", new object[] { tokenOwner, Message.Sender });
+
+        Assert(result.Success, "IsApprovedForAll method call failed.");
+
+        return result.ReturnValue is bool success && success;
+    }
+
+    private void TransferToken(Address contract, ulong tokenId, Address from, Address to)
+    {
+        var result = Call(contract, 0, "TransferFrom", new object[] { from, to, tokenId });
+
+        Assert(result.Success && result.ReturnValue is bool success && success, "The token transfer failed. Be sure contract is approved to transfer token.");
```

Comment:
would be shorter to just check `result.Success && (bool)result.ReturnValue`

Reply:
this can throw casting exception so this is safer